tests/posix/semaphore/src/sem.c

```diff
@@ -54,8 +54,132 @@ static void test_sema(void)
 	zassert_false(sem_destroy(&sema), "sema is not destroyed\n");
 }
 
+/*
+ * Test for named semaphores
+ */
+
+#define N_THRD (3)
+#define SSZ (64)
+
+const char sem_name[] = "/shared_sem";
+int shared_counter = 0;
+
+/*
+ * Flags to track the status of child threads
+ */
+sem_t *ssem_value[N_THRD];
+unsigned int status_flag_open[N_THRD];
+unsigned int status_flag_count[N_THRD];
+unsigned int status_flag_close[N_THRD];
+unsigned int unlinked = 0;
+
+/*
+ * Simple test:
+ * - all child threads receive the same semaphore
+ * - can be used for shared variable access
+ * - can be destroyed properly
+ */
+void *child_code(void *vid)
+{
+	unsigned id = (unsigned) vid;
+	sem_t *ssem;
+	int rv;
+
+	/* Check open */
+	ssem = sem_open(sem_name, O_CREAT, 0666, 1);
+	ssem_value[id] = ssem;
+	status_flag_open[id] = 1;
+
+	/* Check correct locking and counter */
+	rv = sem_wait(ssem);
+	shared_counter++;
+	rv = sem_post(ssem);
+	status_flag_count[id] = 1;
+
+	/* Check close */
+	rv = sem_close(ssem);
+	status_flag_close[id] = 1;
+
+	/* Thread 0 takes care of unlinking */
+	if (id == 0) {
+		rv = sem_unlink(sem_name);
+		unlinked = 1;
+	}
+
+	return NULL;
+}
+
+K_THREAD_DEFINE(t0, SSZ, child_code, (void *)0, NULL, NULL, K_USER,
+		0, K_FOREVER);
+K_THREAD_DEFINE(t1, SSZ, child_code, (void *)1, NULL, NULL, K_USER,
+		0, K_FOREVER);
+K_THREAD_DEFINE(t2, SSZ, child_code, (void *)2, NULL, NULL, K_USER,
+		0, K_FOREVER);
+
+/*
+ * Checks if a status array is all set to 1.
+ */
+static inline bool check_barrier(unsigned int status[N_THRD])
+{
+	for (int i = 0; i < N_THRD; ++i) {
+		if (status[i] == 0)
+			return false;
+	}
+
+	return true;
+}
+
+/*
+ * Checks if the semaphore contains the same value for all threads.
+ */
+static inline bool check_sem(sem_t *sarr[N_THRD])
+{
+	for (int i = 1; i < N_THRD; ++i) {
+		if (sarr[0] != sarr[i])
+			return false;
+	}
+
+	return true;
+}
+
+
+/*
+ * This function behaves as a checker, and taking care all threads pass all
+ * checks.
+ * TODO: add timeout.
+ */
+static void test_named_semaphores(void)
+{
+	/* Launch threads */
+	k_thread_start(t0);
+	k_thread_start(t1);
+	k_thread_start(t2);
```

Comment:
Why are you bothering to create the threads with K_FOREVER timeouts if you're just going to start them synchronously before any initialization happens anyway?  Set the timeout to zero and just remove these three lines and the code won't behave any differently.